### ISSUE TITLE
Tail duplication bug caused by long lines

### ISSUE THREAD

**Issue by jlunavtgrad** [fluent/fluent-bit#3488]
## Bug Report

Using `Skip_Long_Line` with a `DB` can lead fluent-bit to reprocess the same content from a file multiple times.

**To Reproduce**
Step 1: Generate a file with some long lines
```bash
# 50 lines of data
> for i in `seq 1 50`; do openssl rand -hex 100 >> test.log; done
# One very long line
> openssl rand -hex 320000 >> test.log
# 25 more lines of data
> for i in `seq 1 25`; do openssl rand -hex 100 >> test.log; done
```

Step 2: Configure fluent-bit
```
[INPUT]
  Name tail
  Path test.log
  DB fluent-bit.db
  Skip_Long_Lines On
  Read_From_Head On

[OUTPUT]
  Name counter
  Match *
```

Step 3: Run fluent-bit
```bash
> fluent-bit -c fluent-bit.conf
...
1621051660.514340,75 (total = 75)
```
Observe that fluent-bit reads the full file, skipping the long line as expected

Step 4: Restart fluent-bit
Stop fluent-bit. Optionally examine the offset saved in the sqlite database.
```bash
# check the offset in the sqlite db
> sqlite3 fluent-bit.db "select offset from in_tail_files where name = 'test.log'"
32502
# check the file size
stat -c%s test.log
655076
# Restart fluent-bit
> fluent-bit -c fluent-bit.conf
...
1621052184.127856,25 (total = 25)
```
Observe that the file offset in the db is incorrect, and that when fluent-bit is restarted it outputs the last 25 lines of the file a second time.

**Expected behavior**
The expected behavior is for fluent-bit to skip the long line and correctly save the offset in the db such that when the process is restarted it resumes reading from the end of the file.

**Your Environment**

* Version used: 1.7.4
* Operating System and version: linux
* Filters and plugins: tail

**Additional context**
I've been trying to hunt down the cause of duplicates for over a year now, this is my white whale! So glad to finally have a solid repro case!

We use fluent-bit to monitor roughly 1M log files at time and a few of our services have a known habit of occasionally logging more than 100mb of data in a single line. On nodes that have scheduled restarts this leads to thousands of records of duplication sent to our backend.

**Comment by edsiper** [comment]
Assigned to @nokute78

**Comment by nokute78** [comment]
This issue seems to be fixed by v1.7.5. 
@jlunavtgrad  Can you test v1.7.5 ?

## v1.7.5
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.7.5
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/05/15 14:51:41] [ info] [engine] started (pid=87469)
[2021/05/15 14:51:41] [ info] [storage] version=1.1.1, initializing...
[2021/05/15 14:51:41] [ info] [storage] in-memory
[2021/05/15 14:51:41] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/05/15 14:51:41] [ info] [sp] stream processor started
[2021/05/15 14:51:41] [ info] [input:tail:tail.0] inotify_fs_add(): inode=1446937 watch_fd=1 name=test.log
1621057906.773585,76 (total = 76)
^C[2021/05/15 14:51:47] [engine] caught signal (SIGINT)
[2021/05/15 14:51:47] [ info] [input] pausing tail.0
[2021/05/15 14:51:47] [ warn] [engine] service will stop in 5 seconds
[2021/05/15 14:51:51] [ info] [engine] service stopped
[2021/05/15 14:51:51] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=1446937 watch_fd=1
taka@locals:~/git/fluent-bit/build/3488$ sqlite3 fluent-bit.db "select offset from in_tail_files where name = 'test.log'"
655076
```

## v1.7.4
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.7.4
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/05/15 14:49:14] [ info] [engine] started (pid=83076)
[2021/05/15 14:49:14] [ info] [storage] version=1.1.1, initializing...
[2021/05/15 14:49:14] [ info] [storage] in-memory
[2021/05/15 14:49:14] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/05/15 14:49:14] [ info] [sp] stream processor started
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ warn] [input:tail:tail.0] file=test.log have long lines. Skipping long lines.
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input chunk] skip ingesting data with 0 bytes
[2021/05/15 14:49:14] [ info] [input:tail:tail.0] inotify_fs_add(): inode=1446937 watch_fd=1 name=test.log
1621057758.773480,75 (total = 75)
^C[2021/05/15 14:49:21] [engine] caught signal (SIGINT)
[2021/05/15 14:49:21] [ info] [input] pausing tail.0
[2021/05/15 14:49:21] [ warn] [engine] service will stop in 5 seconds
[2021/05/15 14:49:25] [ info] [engine] service stopped
[2021/05/15 14:49:25] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=1446937 watch_fd=1
taka@locals:~/git/fluent-bit/build/3488$ sqlite3 fluent-bit.db "select offset from in_tail_files where name = 'test.log'"
32502
```

**Comment by nokute78** [comment]
Hmm, v1.7.5 ignores `Skip_Long_Lines` ?

In spite of `Skip_Long_Lines On`, v1.7.5 outputs the following log.
```
1621057906.773585,76 (total = 76)
```

**Comment by jlunavtgrad** [comment]
Wow, thanks for the quick response. Version 1.7.5 definitely changes the behavior and won't lead to duplication. Unfortunately it introduces a new, albeit less sever issue. The tail-end of the long line is passed through to the outputs.

The old behavior was to skip these lines entirely, but a nice feature would be to truncate and return the first Buffer_Chunk_Size bytes of the line with a field indicating the line was truncated and the original size.

**Comment by jlunavtgrad** [comment]
I'll close this defect since the newest code acts much differently. I've opened a new defect (#3491) to address the current behavior, and reference the fix for it (#3490) waiting to be merged
